app/core/EngineService/EngineService.ts

```diff
@@ -215,8 +236,61 @@ export class EngineService {
             return;
           }
 
+          // Create debounced persist function (reused for both initial and ongoing persistence)
           const persistController = createPersistController(200);
 
+          // Check if state changed during Engine.init()
+          // Compare 1 level deep - check if any property of the filtered state differs
+          // from the initial state. Controllers preserve referential equality for unchanged properties.
+          // If initialControllerState is undefined (new install or new controller), always persist.
+          // @ts-expect-error - Engine context has stateless controllers
+          const currentState = UntypedEngine.context[controllerName]?.state;
+
+          // Only check for init-time state changes if controller has state
+          // (stateless controllers will have undefined state)
+          if (currentState) {
+            try {
+              const initialControllerState = initialState?.[controllerName] as
+                | Record<string, unknown>
+                | undefined;
+              const filteredState = getPersistentState(
+                currentState,
+                controllerMetadata,
+              );
+
+              // Check if any property at the first level has changed
+              // Only persist if there's state to persist AND either:
+              // 1. No initial state existed but now there is state (new install/new controller)
+              // 2. Initial state existed and has changed
+              const hasStateToSave = Object.keys(filteredState).length > 0;
+              const hasChanged =
+                hasStateToSave &&
+                (!initialControllerState ||
+                  Object.keys(filteredState).some(
+                    (key) =>
+                      filteredState[key] !==
+                      initialControllerState[
+                        key as keyof typeof initialControllerState
+                      ],
```

Comment:
SonarCloud seems to be correct here, the type assertion is not needed

```suggestion
                    (key) => filteredState[key] !== initialControllerState[key],
```